envoy/upstream/outlier_detection.h

```diff
@@ -47,6 +49,104 @@ enum class Result {
   ExtOriginRequestSuccess // Request was completed successfully.
 };
 
+// Types of errors reported to outlier detectors.
+// Each type may have a different syntax and content.
+// TODO (cpakulski) - maybe use enums from protobufs.
+enum class ExtResultType {
+  TEST, // Used in unit tests.
+  HTTP_CODE,
+  LOCAL_ORIGIN,
+};
+
+/*
+ * Class carries result of a transaction with upstream entity
+ * or generated internally by Envoy.
+ * Different categories of results will be derived from that base class.
+ * Those categories of results are fed only into Outlier Detection extensions.
+ */
+class ExtResult {
+public:
+  ExtResult() = delete;
+  ExtResult(ExtResultType type) : type_(type) {}
+  virtual ExtResultType type() const { return type_; };
+  virtual ~ExtResult() = default;
+
+private:
+  const ExtResultType type_;
+};
+
+// Types derived from TypedExtResult are used to report the result of transaction to
+// an outlier detection monitor.
+template <ExtResultType E> class TypedExtResult : public ExtResult {
+protected:
+  TypedExtResult() : ExtResult(E) {}
+};
+
+class HttpCode : public TypedExtResult<ExtResultType::HTTP_CODE> {
+public:
+  HttpCode(uint32_t code) : code_(code) {}
+  HttpCode() = delete;
+  virtual ~HttpCode() {}
+  uint32_t code() const { return code_; }
+
+private:
+  uint32_t code_;
+};
+
+// LocalOriginEvent is used to report errors like resets, timeouts but also
+// successful connection attempts.
+class LocalOriginEvent : public TypedExtResult<ExtResultType::LOCAL_ORIGIN> {
+public:
+  LocalOriginEvent(Result result) : result_(result) {}
+  LocalOriginEvent() = delete;
+  Result result() const { return result_; }
+
+private:
+  Result result_;
+};
+
+// Base class for various types of monitors.
+// Each monitor may implement different health detection algorithm.
+class ExtMonitor {
+public:
+  ExtMonitor(const std::string& name, uint32_t enforce) : name_(name), enforce_(enforce) {}
+  ExtMonitor() = delete;
+  virtual ~ExtMonitor() {}
+  virtual void reportResult(const ExtResult&) PURE;
+
+  void
+  setCallback(std::function<void(uint32_t, std::string, absl::optional<std::string>)> callback) {
+    callback_ = callback;
+  }
+
+  void reset() { onReset(); }
+  std::string name() const { return name_; }
+
+protected:
+  virtual bool onError() PURE;
+  virtual void onSuccess() PURE;
+  virtual void onReset() PURE;
+  virtual std::string getFailedExtraInfo() { return ""; }
+
+  std::string name_;
+  uint32_t enforce_{100};
+  std::function<void(uint32_t, std::string, absl::optional<std::string>)> callback_;
+};
```

Comment:
I think only `void reportResult()` and `setCallbacks()` is necessary here as the pure abstract interface of `ExtMonitor`. All others things should be moved to `ExtMonitorBase` or something.

And `void putResult()` and `void setExtMonitorCallbacks()` name maybe better.

And note, rather than provide a lambda as callbacks, please define a new abstract class as the callbacks, like `class ExtMonitorCallbacks`.
By this way, we can add more methods to the callbacks in the future.

Reply:
Thanks for catching it! There was already `ExtMonitorBase` class, so the move was simple. Simplified the pure abstract class ExtMonitor to contain only `putResult`, `setExtMonitorCallback` and `reset`.

Reply:
WDYT about:

> And note, rather than provide a lambda as callbacks, please define a new abstract class as the callbacks, like class ExtMonitorCallbacks. By this way, we can add more methods to the callbacks in the future.

Reply:
To be honest, I do not think that new methods will be added. In fact, the opposite is true. I hope to remove that callback when "old" outlier is deprecated. The reason I had to add it is that "enforce" parameter is taken via the following global function: https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/outlier_detection_impl.cc#L393-L420, which works great for "canned/static" outlier monitors, but a new solution is dynamic so I had to bridge it via callback. 
But I can add `ExtMonitorCallback` if you prefer. Let me know.

Reply:
>  I hope to remove that callback when "old" outlier is deprecated.

This may never happen except we can find a way that doesn't effect exist users.

I am not pretty sure which would be best way. So, let the time to tell. Please give the lambda an alias name `ExtMonitorCallback` first. And if new methods are necessary, we can change it in new PR. :)

Reply:
Done. Added a new type ExtMonitorCallback.